IPython/kernel/comm/comm.py

```diff
@@ -63,6 +66,10 @@ def __init__(self, target_name='', data=None, **kwargs):
     
     def _publish_msg(self, msg_type, data=None, metadata=None, buffers=None, **keys):
         """Helper for sending a comm message on IOPub"""
+        if threading.current_thread().name != 'MainThread' and IOLoop.initialized():
+            # make sure we never send on a zmq socket outside the main IOLoop thread
+            IOLoop.instance().add_callback(lambda : self._publish_msg(msg_type, data, metadata, buffers, **keys))
```

Comment:
all callbacks fire each eventloop iteration. For threadsafety reasons, it's more like:

```python
with callback_lock:
    new_callbacks = []
    callbacks = self.callbacks
    self.callbacks = new_callbacks
for cb in callbacks:
    cb() # can schedule new callbacks, which won't fire until the next iteration
```

I don't know what you mean by overflow since it's a Python list, but I don't think there's any problem.